File: hedera-mirror-rosetta/test/mocks/transaction_repository.go
/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

package mocks

import (
	"context"

	rTypes "github.com/coinbase/rosetta-sdk-go/types"
	"github.com/hashgraph/hedera-mirror-node/hedera-mirror-rosetta/app/domain/types"
	"github.com/stretchr/testify/mock"
)

var NilTransaction *types.Transaction

type MockTransactionRepository struct {
	mock.Mock
}

func (m *MockTransactionRepository) FindByHashInBlock(
	ctx context.Context,
	identifier string,
	consensusStart int64,
	consensusEnd int64,
) (*types.Transaction, *rTypes.Error) {
	args := m.Called()
	return args.Get(0).(*types.Transaction), args.Get(1).(*rTypes.Error)
}

func (m *MockTransactionRepository) FindBetween(ctx context.Context, start, end int64) (
	[]*types.Transaction,
	*rTypes.Error,
) {
	args := m.Called()
	return args.Get(0).([]*types.Transaction), args.Get(1).(*rTypes.Error)
}

func (m *MockTransactionRepository) Types(ctx context.Context) (map[int]string, *rTypes.Error) {
	panic("implement me")
}

func (m *MockTransactionRepository) TypesAsArray(ctx context.Context) ([]string, *rTypes.Error) {
	args := m.Called()
	return args.Get(0).([]string), args.Get(1).(*rTypes.Error)
}

func (m *MockTransactionRepository) Results(ctx context.Context) (map[int]string, *rTypes.Error) {
	args := m.Called()
	return args.Get(0).(map[int]string), args.Get(1).(*rTypes.Error)
}

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TransactionRepository.java
/*
 * Copyright (C) 2024 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.mirror.web3.repository;

import com.hedera.mirror.common.domain.entity.EntityId;
import com.hedera.mirror.common.domain.transaction.Transaction;
import java.util.Optional;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.CrudRepository;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

@Repository
public interface TransactionRepository extends CrudRepository<Transaction, Long> {

    @Query("select t from Transaction t " +
            "where t.payerAccountId = :payerAccountId " +
            "and t.validStartNs = :validStartNs " +
            "and t.consensusTimestamp >= :validStartNs and t.consensusTimestamp <= :maxConsensusNs")
    Optional<Transaction> findByPayerAccountIdAndValidStartNsAndConsensusTimestampBefore(
            @Param("payerAccountId") EntityId payerAccountId,
            @Param("validStartNs") long validStartNs,
            @Param("maxConsensusNs") long maxConsensusNs);
}